src/cpu/o3/tdm.cc

```diff
@@ -0,0 +1,260 @@
+#include "cpu/o3/cpu.hh"
+#include "cpu/o3/decode.hh"
+#include "cpu/o3/fetch.hh"
+#include "cpu/o3/iew.hh"
+#include "cpu/o3/rename.hh"
+
+namespace gem5
+{
+
+namespace o3
+{
+TopDownStats::TopDownStats(CPU *cpu, Fetch *fetch, Rename *rename,
+                           Decode *decode, IEW *iew, Commit *commit)
+    : statistics::Group(cpu, "TopDownStats"),
+      topDownL1(cpu, fetch, rename, decode, iew, commit),
+      topDownFbL2(cpu, fetch),
+      topDownBsL2(cpu, decode, iew),
+      topDownBbL2(cpu, rename, iew),
+      topDownBbMem(cpu, rename, iew)
+{}
+
+TopDownStats::TopDownL1::TopDownL1(CPU *cpu, Fetch *fetch, Rename *rename,
+                                   Decode *decode, IEW *iew, Commit *commit)
+    : statistics::Group(cpu, "TopDownL1"),
+      ADD_STAT(frontendBound,
+               statistics::units::Rate<statistics::units::Count,
+                                       statistics::units::Count>::get(),
+               "Frontend Bound, fraction of slots lost due to frontend "
+               "undersupplying the backend"),
+      ADD_STAT(badSpeculation,
+               statistics::units::Rate<statistics::units::Count,
+                                       statistics::units::Count>::get(),
+               "Bad Speculation, fraction of slots lost due to mispeculation"),
+      ADD_STAT(backendBound,
+               statistics::units::Rate<statistics::units::Count,
+                                       statistics::units::Count>::get(),
+               "Backend Bound, fraction of slots lost due to backend resource"
+               " constraints."),
+      ADD_STAT(
+          retiring,
+          statistics::units::Rate<statistics::units::Count,
+                                  statistics::units::Count>::get(),
+          "Retiring, fraction of slots successfully retired by the backend")
+{
+    // Total Slots
+    statistics::Temp totalSlots =
+        rename->getWidth() * cpu->baseStats.numCycles;
+
+    // L1 Frontend Bound
+    frontendBound = fetch->getStats().fetchBubbles / (totalSlots);
+
+    // L1 Bad Speculation
+    // Recovery cycles for mispredictions detected at Decode
+    int recoveryCycleToDecode = decode->getFetchToDecodeDelay();
+
+    auto decodeBranchMispred =
+        (int)recoveryCycleToDecode * decode->getStats().branchMispred;
+
+    // Recovery cycles for mispredictions detected at IEW
+    int recoveryCycleToIEW = decode->getFetchToDecodeDelay() +
+                             rename->getDecodeToRenameDelay() +
+                             iew->getRenameToIEWDelay();
+
+    auto iewBadSpec =
+        (int)recoveryCycleToIEW * (iew->getStats().branchMispredicts +
+                                   iew->getStats().memOrderViolationEvents);
+
+    // Number of wasted slots due to bad speculation
+    auto wastedSlots =
+        rename->getStats().renamedInsts - commit->getStats().committedInst;
+
+    badSpeculation = (wastedSlots + (decodeBranchMispred + iewBadSpec) *
```

Comment:
I don't think `decodeBranchMispred` count to badSpeculation. These are BTB misses and are part of the frontend-bound fraction.